src/Magnum/Octree/Octree.h

```diff
@@ -0,0 +1,274 @@
+        Int depth() const {
+            return _depth;
+        }
+
+        /** @brief Get the bounding boxes in this octree node */
+        const std::vector<Range3D>& entries() const {
+            return _entries;
+        }
+
+        /** @brief Get the data contained in this octree node */
+        const std::vector<T>& data() const {
+            return _data;
+        }
+
+        /** @brief Get the i-th child of this octree */
+        Octree<T>* child(int i) const {
+            return _children[i];
+        }
+
+        /**
+         * @brief Check if the octree is a leaf node.
+         *
+         * An octree is a leaf node if it has no children.
+         */
+        bool isLeafNode() const {
+            return _isLeafNode;
+        }
+
+        /**
+         * @brief Set if octree is a leaf node.
+         *
+         * An octree is a leaf node if it has no children.
+         */
+        void setLeafNode(bool isLeafNode) {
+            _isLeafNode = isLeafNode;
+        }
+
+        /** @brief Whether this node is empty */
+        bool empty() const {
+            return _data.size() == 0;
+        }
+
+        /**
+         * @brief Build the octree for points and a maximal depth
+         * @param points      The points that should get stored in the octree
+         * @param maxDepth    The maximal depth of the octree
+         * @return Reference to self (for method chaining)
+         */
+        Octree<T>& build(const std::vector<Range3D>& entries, const std::vector<T>& data, const Int maxDepth) {
+            CORRADE_ASSERT(data.size() == entries.size(), "Entries and data need to be same size.", *this);
+
+            auto entryItor = entries.begin();
+            auto entryEnd = entries.end();
+            auto dataItor = data.begin();
+
+            createInitialBounds(entries, maxDepth);
+
+            while(entryItor != entryEnd) {
+                insert(*entryItor, *dataItor);
+                entryItor++;
+                dataItor++;
+            }
+
+            return *this;
+        }
+
+        /**
+         * @brief Get the child of the octree in which the point is stored, if the point exists
+         * @param point    The point that should be found
+         * @return The index of the child potentially containing the point
+         */
+        Int getOctantContainingBox(const Range3D& point) const;
+
+        /**
+         * @brief Get all points contained of the octree within given bounds
+         * @param found    All the points found in the specified bounds
+         * @param min      The minimal corner of the search bounds
+         * @param max      The maximal corner of the search bounds
+         * @return Reference to self (for method chaining)
+         */
+        Octree<T>& points(std::vector<T>& resultData, const Vector3& min, const Vector3& max);
+        Octree<T>& points(std::vector<T>& resultData, const Vector4 (&frustum)[6]);
+        Octree<T>& points(std::vector<T>& resultData);

```

Comment:
I don't see any value in method chaining for getters. Could this be rewritten as `std::vector<T> points(...)`? Then, later, possibly with passing the storage as optional rvalue to avoid unnecessary reallocations.